Update deployment workflow for final HSM solution

andyfeller opened the pull request:
fixes https://github.com/github/cli/issues/213

This applies the changes from the separate Windows HSM signing prototype development to the official deployment workflow including:

1. Use of Azure Code Signing client
2. Sourcing signtool.exe from runner
3. Moving from batch to PowerShell for Windows signing script
4. Using the same signing process for .exe and .msi

andyfeller committed:
Update deployment workflow for final HSM solution

This applies the changes from the separate Windows HSM signing prototype development to the official deployment workflow including:

1. Use of Azure Code Signing client
2. Sourcing signtool.exe from runner
3. Moving from batch to PowerShell for Windows signing script
4. Using the same signing process for .exe and .msi

andyfeller commented:
@samcoe @williammartin @mxie @niik: these changes are based on the prototyping done in the now deleted `.github/workflows/deployment-hsm.yml` workflow with https://github.com/cli/cli/actions/runs/7211896336/job/19648372331 being the latest successful workflow

andyfeller commented on .goreleaser.yml:
```diff
@@ -41,7 +41,7 @@ builds:
       post:
         - cmd: >-
-            {{ if eq .Runtime.Goos "windows" }}.\script\sign{{ else }}./script/sign{{ end }} '{{ .Path }}'
+            {{ if eq .Runtime.Goos "windows" }}pwsh .\script\sign.ps1{{ else }}./script/sign{{ end }} '{{ .Path }}'
```
Calling out that `script/sign.ps1` is carried over from the prototyping as-is, so don't be alarmed by all of the `script/sign*` deletion in the PR 😅

williammartin reviewed (commented):
Do you think we should have a step where we upload build artifacts so that we can check the results of signing before approving the `release` step? Or do you feel confident that the windows job would fail and not continue to the release step if there was an issue?

andyfeller commented:
> Do you think we should have a step where we upload build artifacts so that we can check the results of signing before approving the `release` step? Or do you feel confident that the windows job would fail and not continue to the release step if there was an issue?

Oooo, good question.  I hesitate to permanently change the deployment process to require 2 approvals unless all of the core maintainers were onboard given that it is like to forget the 2nd approval.  That said, let me try testing something out with an alternate branch with the prototype workflow to see how it handles failure signing and failing the workflow

andyfeller committed:
Ensure Windows sign script fails based on signtool

This is a change based on PR discussion about ensuring the sign script fails if signtool errors, passing the last exit code from the called command as the result of the script.

For more information, see https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4#lastexitcode

andyfeller commented:
> > Do you think we should have a step where we upload build artifacts so that we can check the results of signing before approving the `release` step? Or do you feel confident that the windows job would fail and not continue to the release step if there was an issue?
> 
> Oooo, good question. I hesitate to permanently change the deployment process to require 2 approvals unless all of the core maintainers were onboard given that it is like to forget the 2nd approval. That said, let me try testing something out with an alternate branch with the prototype workflow to see how it handles failure signing and failing the workflow

Testing out an enhancement in https://github.com/cli/cli/actions/runs/7224626745/job/19686310726, I just pushed a change to ensure that `script/sign.ps1` will exit with the exit code from the signtool child process.  This is the best I think we can do for now, iterating as we discover new fun things about our signtool friend

With this, I'm confident these are the changes we need, so I've enabled auto-merge to merge this once all the checks pass.
